src/utils/pgwire/src/pg_extended.rs

```diff
@@ -20,20 +20,42 @@ use regex::Regex;
 use crate::pg_field_descriptor::{PgFieldDescriptor, TypeOid};
 use crate::pg_protocol::cstr_to_str;
 
+/// Replace generic params in query into real params.
+///
+/// # Example
+///
+/// ```
+/// let raw_params = vec!["A".into(), "B".into(), "C".into()];
+/// let type_description = vec![TypeOid::Varchar; 3];
+/// let params = parse_params(&type_description, &raw_params);
+/// let res = replace_params("SELECT $3,$2,$1".to_string(), &[1, 2, 3], &params);
+/// assert_eq!(res, "SELECT \'C\',\'B\',\'A\'");
+/// ```
 fn replace_params(query_string: String, generic_params: &[usize], params: &[String]) -> String {
     let mut tmp = query_string;
-    tmp.push(' ');
-    for &i in generic_params.iter() {
-        let pattern = Regex::new(format!(r"(?P<x>\${})(?P<y>[,;\s]+)", i).as_str()).unwrap();
-        let param = &params[i.sub(1)];
+
+    for &param_idx in generic_params {
+        let pattern =
+            Regex::new(format!(r"(?P<x>\${0})(?P<y>[,;\s]+)|\${0}$", param_idx).as_str()).unwrap();
+        let param = &params[param_idx.sub(1)];
         tmp = pattern
             .replace_all(&tmp, format!("{}$y", param))
             .to_string();
     }
-    tmp.pop();
+
     tmp
 }
 
+/// Parse params accoring the type description.
+///
+/// # Example
+///
+/// ```
+/// let raw_params = vec!["A".into(), "B".into(), "C".into()];
+/// let type_description = vec![TypeOid::Varchar; 3];
+/// let params = parse_params(&type_description, &raw_params);
+/// assert_eq!(params, vec!["\'A\'", "\'B\'", "\'C\'"])
```

Comment:
what's the usage of `\ `?

Reply:
I want to represent escape character, but I find I can directly use ' , I will fix it later.